### PR TITLE
Feat/4064  notes translation

### PR DESCRIPTION
**Pull request by asf-algoo**
See #4064.

## Checkpoints

**For developers**

- [x] If relevant, manual tests have been done to ensure the stability of the whole application and that the involved feature works
- [x] The original issue is up to date w.r.t the latest discussions and contains a short summary of the implemented solution
- [x] Automated tests covering the feature or the fix, have been written, deemed irrelevant (give the reason), or an issue has been created to implement the test (give the link)
- [x] Make sure that:
  - if there are modifications in the Tracim configuration files (eg. `development.ini`), they are documented in `backend/doc/setting.md`
  - any migration process required for existing instances is documented
  - relevant people for these changes are notified
- [x] Original authors of the features included in a multi-feature branch (maintenance fixes -> develop, security fixes -> develop, …) should be part of the reviewers, especially if you encountered merge conflicts.

**For code reviewers**

- [x] The code is clear enough
- [x] If there are FIXMEs in the code, related issues are mentioned in the FIXME
- [x] If there are TODOs, NOTEs or HACKs in code, the date and the developer initials are present

**For testers**

- [x] Manual, quality tests have been done

**Testing method**

Launch the dev backend with `run_dev_backend.sh`.
Create a note and translate it with the *Show translation* button. The "translation" is a placeholder table describing the request's input (this is for tests and is normal).

### CODE REVIEW

File: frontend_app_html-document/i18next.scanner/pt/translation.json [hunk 1]
```diff
@@ -29,5 +29,11 @@
   "The content has been modified by {{author}}": "O conteúdo foi modificado por {{author}}",
   "Error while saving the user configuration": "Erro ao guardar a configuração do utilizador",
   "Your text contains mentions that do not match any member of this space:": "O seu texto contém menções que não correspondem a nenhum membro deste espaço:",
-  "Validate anyway": "Validar de qualquer forma"
+  "Validate anyway": "Validar de qualquer forma",
+  "Restore the original language": "Restaurar a língua original",
+  "Show translation": "Visualizar tradução",
+  "Translation pending…": "__NOT_TRANSLATED__",
```
Comment (gbonaspetti) [inline frontend_app_html-document/i18next.scanner/pt/translation.json:35]:
```suggestion
  "Translation pending…": "Tradução pendente...",
```

File: frontend_app_html-document/i18next.scanner/pt/translation.json [hunk 2]
```diff
@@ -29,5 +29,11 @@
...
+  "Translation service is not available": "__NOT_TRANSLATED__",
```
Comment (gbonaspetti) [inline frontend_app_html-document/i18next.scanner/pt/translation.json:36]:
```suggestion
  "Translation service is not available": "O serviço de tradução não está disponível",
```

File: frontend_app_html-document/i18next.scanner/pt/translation.json [hunk 3]
```diff
@@ -29,5 +29,11 @@
...
+  "Unknown error": "__NOT_TRANSLATED__",
```
Comment (gbonaspetti) [inline frontend_app_html-document/i18next.scanner/pt/translation.json:37]:
```suggestion
  "Unknown error": "Erro desconhecido",
```

File: frontend_app_html-document/i18next.scanner/pt/translation.json [hunk 4]
```diff
@@ -29,5 +29,11 @@
...
+  "Error while translating the document: {{details}}": "__NOT_TRANSLATED__"
```
Comment (gbonaspetti) [inline frontend_app_html-document/i18next.scanner/pt/translation.json:38]:
```suggestion
  "Error while translating the document: {{details}}": "Erro ao traduzir o documento:"
```

File: frontend_app_html-document/src/container/HtmlDocument.jsx [hunk 5]
```diff
@@ -90,7 +94,10 @@ export class HtmlDocument extends React.Component {
       invalidMentionList: [],
       oldInvalidMentionList: [],
       showInvalidMentionPopupInComment: false,
-      showInvalidMentionPopupInContent: false
+      showInvalidMentionPopupInContent: false,
+      translatedRawContent: null,
+      defaultTranslationState: defaultTranslationState,
```
Comment (raphj) [inline frontend_app_html-document/src/container/HtmlDocument.jsx:99]: It seems this is never changed, maybe this should be kept away from the state?

**Review by raphj — APPROVED** [review]

File: frontend_app_html-document/src/container/HtmlDocument.jsx (hunk 5 above)
Comment (grignards) [reply frontend_app_html-document/src/container/HtmlDocument.jsx:99]: I hesitated, theoretically it could change as the default value comes from a property.
And it is unsure where to keep this if it is not in the state: here I could put it in the component's object but wouldn't work with a pure component (and later on with hooks).

File: frontend_app_html-document/src/container/HtmlDocument.jsx [hunk 6]
```diff
@@ -842,15 +908,15 @@ export class HtmlDocument extends React.Component {
...
             onChangeText={this.handleChangeText}
             isArchived={state.content.is_archived}
             isDeleted={state.content.is_deleted}
             isDeprecated={state.content.status === state.config.availableStatuses[3].slug}
             deprecatedStatus={state.config.availableStatuses[3]}
             isDraftAvailable={state.mode === APP_FEATURE_MODE.VIEW && state.loggedUser.userRoleIdInWorkspace >= ROLE.contributor.id && getLocalStorageItem(state.appName, state.content, LOCAL_STORAGE_FIELD.RAW_CONTENT)}
-            onClickRestoreArchived={this.handleClickRestoreArchive}
-            onClickRestoreDeleted={this.handleClickRestoreDelete}
+            // onClickRestoreArchived={this.handleClickRestoreArchive}
+            // onClickRestoreDeleted={this.handleClickRestoreDelete}
```
Comment (Skylsmoi) [inline frontend_app_html-document/src/container/HtmlDocument.jsx:919]: onClickRestoreArchived is indeed unused but onClickRestoreDeleted is and should be kept

File: frontend_lib/src/component/Button/IconButton.jsx [hunk 7]
```diff
@@ -30,7 +30,7 @@ const IconButton = props => {
       data-cy={props.dataCy}
     >
       {props.icon && <i className={`fa-fw ${props.icon} iconbutton__icon`} />}
-      {props.text && <span className='iconbutton__text'>{props.text}</span>}
+      {props.text && <span className={props.icon ? 'iconbutton__text_with_icon' :'iconbutton__text'}>{props.text}</span>}
```
Comment (Skylsmoi) [inline frontend_lib/src/component/Button/IconButton.jsx:33]: missing space before `'iconbutton__text'`, I think standard would whine about it

File: frontend_lib/src/component/Button/IconButton.styl [hunk 8]
```diff
@@ -41,10 +40,17 @@
...
     &:hover
       text-decoration underline
+    &_light
+      color mediumDarkGrey
+      &:hover
+        color darkGrey
+    &_dark
+      color fontColor
+      &:hover
+        color black
```
Comment (Skylsmoi) [inline frontend_lib/src/component/Button/IconButton.styl:53]: in the BOM norm, each "children" should be separated by 2 underscores.
If _light and _dark and modifiers of the class, then the norm expect a dash as a separation char
```suggestion
    &-light
      color mediumDarkGrey
      &:hover
        color darkGrey
    &-dark
      color fontColor
      &:hover
        color black
```

**Review by Skylsmoi — CHANGES_REQUESTED** [review]

File: frontend_app_html-document/src/component/HtmlDocument.jsx [hunk 9]
```diff
@@ -69,25 +89,41 @@ export const HtmlDocument = props => {
...
-              {props.t(
-                'Version #{{versionNumber}}', {
-                  versionNumber: props.mode === APP_FEATURE_MODE.VIEW && !props.isRefreshNeeded
-                    ? props.lastVersion
-                    : props.version
-                }
+            <div className='html-document__contentpage__textnote__top'>
+              {props.translationState === TRANSLATION_STATE.PENDING && (
+                <span className='html-document__contentpage__textnote__top__translation'>
+                  <i className='fa fa-spinner fa-spin' /> {props.t('Translation pending…')}
```
Comment (gbonaspetti) [inline frontend_app_html-document/src/component/HtmlDocument.jsx:97]: Why `props.t('Translation pending…')` is not at `TOGGLE_TRANSLATION_TEXT[TRANSLATION_STATE.PENDING]` and here you can use `TOGGLE_TRANSLATION_TEXT[props.translationState]` ?

File: frontend_app_html-document/src/container/HtmlDocument.jsx [hunk 10]
```diff
@@ -61,7 +62,10 @@ export class HtmlDocument extends React.Component {
 
     const param = props.data || debug
     props.setApiUrl(param.config.apiUrl)
-
+    console.log(`translation config`, param.config.system.config.translation_service__enabled)
```
Comment (gbonaspetti) [inline frontend_app_html-document/src/container/HtmlDocument.jsx:65]: forgotten console

File: frontend_app_html-document/src/container/HtmlDocument.jsx [hunk 11]
```diff
@@ -842,14 +915,14 @@ export class HtmlDocument extends React.Component {
...
-            text={state.content.raw_content}
+            text={displayTranslatedText ? state.translatedRawContent : state.content.raw_content}
             onChangeText={this.handleChangeText}
             isArchived={state.content.is_archived}
             isDeleted={state.content.is_deleted}
             isDeprecated={state.content.status === state.config.availableStatuses[3].slug}
             deprecatedStatus={state.config.availableStatuses[3]}
             isDraftAvailable={state.mode === APP_FEATURE_MODE.VIEW && state.loggedUser.userRoleIdInWorkspace >= ROLE.contributor.id && getLocalStorageItem(state.appName, state.content, LOCAL_STORAGE_FIELD.RAW_CONTENT)}
-            onClickRestoreArchived={this.handleClickRestoreArchive}
+            // onClickRestoreArchived={this.handleClickRestoreArchive}
```
Comment (gbonaspetti) [inline frontend_app_html-document/src/container/HtmlDocument.jsx:925]: forgotten comment

File: frontend_lib/src/component/Button/IconButton.styl [hunk 12]
```diff
@@ -13,49 +12,56 @@
     opacity 0.65
     cursor not-allowed
 
-  &__dark
+  &-dark
```
Comment (gbonaspetti) [inline frontend_lib/src/component/Button/IconButton.styl:15]: I don't understand why this change, can you explain it to me?

File: functionnal_tests/cypress/integration/app_html-document/translate_document.spec.js [hunk 13]
```diff
@@ -0,0 +1,31 @@
+import { PAGES as p } from '../../support/urls_commands.js'
+
+describe('Note/html document translation', function () {
+  before(() => {
+    const workspaceId = 1
+    let contentId = 1
+    cy.resetDB()
+    cy.setupBaseDB()
+    cy.loginAs('administrators')
+    cy.createHtmlDocument("A Note", workspaceId).then(doc => {
```
Comment (gbonaspetti) [inline functionnal_tests/cypress/integration/app_html-document/translate_document.spec.js:10]:
```suggestion
    cy.createHtmlDocument('A Note', workspaceId).then(doc => {
```

File: run_dev_backend.sh [hunk 14]
```diff
@@ -126,6 +126,8 @@ if [ "$mode" = "cypress" ]; then
     tracimcli db delete --force || true
     tracimcli db init || true
     cp /tmp/${DATABASE_NAME}.sqlite /tmp/${DATABASE_NAME}.sqlite.tmp
+    export TRACIM_TRANSLATION_SERVICE__ENABLED=True
+    export TRACIM_TRANSLATION_SERVICE__PROVIDER=test
```
Comment (gbonaspetti) [inline run_dev_backend.sh:130]: This `test`is a forgot debug?

**Review by gbonaspetti — CHANGES_REQUESTED** [review]

File: frontend_app_html-document/src/container/HtmlDocument.jsx (hunk 10 above)
Comment (grignards) [reply frontend_app_html-document/src/container/HtmlDocument.jsx]: Yes, I just removed it.

File: frontend_app_html-document/src/container/HtmlDocument.jsx (hunk 11 above)
Comment (grignards) [reply frontend_app_html-document/src/container/HtmlDocument.jsx:924]: Nope, this is in fact unused (`handleClickRestoreArchive` is already commented and my IDE was complaining that the function did not exist).

File: frontend_lib/src/component/Button/IconButton.styl (hunk 12 above)
Comment (grignards) [reply frontend_lib/src/component/Button/IconButton.styl:15]: See @Skylsmoi comment: https://github.com/tracim/tracim/pull/4220#discussion_r583566302

File: run_dev_backend.sh (hunk 14 above)
Comment (grignards) [reply run_dev_backend.sh:130]: No, this is needed to configure a mock/test-only translation service in the backend so that the cypress tests is functional.
Without it we cannot test the translation in frontend as the translate button is only displayed when the backend has a configured translation service.

File: frontend_app_html-document/src/component/HtmlDocument.jsx (hunk 9 above)
Comment (grignards) [reply frontend_app_html-document/src/component/HtmlDocument.jsx:97]: `toggleTranslationText` is used for knowing which text to put on the translation *button*. I extended it to `null` for PENDING/DISABLED tests mainly for convenience.
This text is only a loader and not (directly) related to the button allowing to translate.

File: frontend_app_html-document/src/container/HtmlDocument.jsx [hunk 15]
```diff
@@ -420,7 +427,9 @@ export class HtmlDocument extends React.Component {
       newComment: localStorageComment || '',
       rawContentBeforeEdit: rawContentBeforeEdit,
       timeline: revisionWithComment,
-      isLastTimelineItemCurrentToken: false
+      isLastTimelineItemCurrentToken: false,
+      translationState: state.defaultTranslationState,
```
[the remark this answers is not in the document]
Comment (grignards) [reply frontend_app_html-document/src/container/HtmlDocument.jsx]: Done.

File: frontend_app_html-document/src/container/HtmlDocument.jsx [hunk 16]
```diff
@@ -532,7 +541,9 @@ export class HtmlDocument extends React.Component {
             raw_content: newDocumentForApiWithMention
           },
           oldInvalidMentionList: allInvalidMentionList,
-          showInvalidMentionPopupInContent: false
+          showInvalidMentionPopupInContent: false,
+          translatedRawContent: null,
+          translationState: state.defaultTranslationState
```
[the remark this answers is not in the document]
Comment (grignards) [reply frontend_app_html-document/src/container/HtmlDocument.jsx]: Done.

File: frontend_app_html-document/src/container/HtmlDocument.jsx [hunk 17]
```diff
@@ -658,9 +669,12 @@ export class HtmlDocument extends React.Component {
         raw_content: revision.raw_content,
         number: revision.number,
         status: revision.status,
+        current_revision_id: revision.revision_id,
         is_archived: prev.is_archived, // archived and delete should always be taken from last version
         is_deleted: prev.is_deleted
       },
+      translationState: state.defaultTranslationState,
```
[the remark this answers is not in the document]
Comment (grignards) [reply frontend_app_html-document/src/container/HtmlDocument.jsx]: Done.

File: frontend_app_html-document/src/container/HtmlDocument.jsx [hunk 18]
```diff
@@ -689,7 +703,9 @@ export class HtmlDocument extends React.Component {
       content: newObjectContent,
       timeline: prev.timeline.map(timelineItem => ({ ...timelineItem, hasBeenRead: true })),
       mode: APP_FEATURE_MODE.VIEW,
-      showRefreshWarning: false
+      showRefreshWarning: false,
+      translatedRawContent: null,
+      translationState: state.defaultTranslationState
```
[the remark this answers is not in the document]
Comment (grignards) [reply frontend_app_html-document/src/container/HtmlDocument.jsx]: Done.

File: frontend_app_html-document/src/container/HtmlDocument.jsx [hunk 19]
```diff
@@ -743,11 +759,61 @@ export class HtmlDocument extends React.Component {
     return !!state.loggedUser.config[`content.${state.content.content_id}.notify_all_members_message`]
   }
 
+  handleTranslateDocument = async () => {
+    const { props, state } = this
+    this.setState({ translationState: TRANSLATION_STATE.PENDING })
+    await new Promise(resolve => {
+      setTimeout(() => resolve(), 500)
+    })
```
[the remark this answers is not in the document]
Comment (grignards) [reply frontend_app_html-document/src/container/HtmlDocument.jsx]: Removed.

File: frontend_app_html-document/src/container/HtmlDocument.jsx (hunk 6 above)
Comment (grignards) [reply frontend_app_html-document/src/container/HtmlDocument.jsx]: Corrected.

File: frontend_lib/src/component/Button/IconButton.jsx (hunk 7 above)
Comment (grignards) [reply frontend_lib/src/component/Button/IconButton.jsx]: Corrected.

File: frontend_lib/src/component/Button/IconButton.styl (hunk 8 above)
Comment (grignards) [reply frontend_lib/src/component/Button/IconButton.styl]: Updated.

**Review by Skylsmoi — APPROVED** [review]

**Review by gbonaspetti — APPROVED** [review]
